Documentation: Matching Utilities - isAnyOf, isAllOf (#806)
reduxjs/redux-toolkit: docs/api/matching-utilities.mdx

---
id: matching-utilities
title: Matching Utilities
sidebar_label: Matching Utilities
hide_title: true
---

# Matching Utilities

> Redux Toolkit exports several type-safe utility matchers that you can leverage to simplify your code.

Reviewer: ```suggestion
Redux Toolkit exports several type-safe action matching utilities that you can leverage when checking for specific kinds of actions.  These are primarily useful for the `builder.addMatcher()` cases in `createSlice` and `createReducer`, as well as when writing custom middleware.
```

### General Purpose

- [`isAllOf`](#isallof) - returns true when **all** conditions are met

> - [`isAnyOf`](#isanyof) - returns true when at least one of the conditions are met

Reviewer: ```suggestion
- [`isAnyOf`](#isanyof) - returns true when **at least one of** the conditions are met
```

```ts title="Example without using a matcher utility"
import {
  createAsyncThunk,
  createReducer,
  PayloadAction
} from '@reduxjs/toolkit'

interface Data {
  isInteresting: boolean
  isSpecial: boolean
}

interface Special extends Data {
  isSpecial: true
}

interface Interesting extends Data {
  isInteresting: true
}

function isSpecial(
  action: PayloadAction<Data>
): action is PayloadAction<Special> {
  return action.payload.isSpecial
}

function isInteresting(
  action: PayloadAction<Data>
): action is PayloadAction<Interesting> {
  return action.payload.isInteresting
}

interface ExampleState {
  isSpecial: boolean
  isInteresting: boolean
}

const initialState: ExampleState = {
  isSpecial: false,
  isInteresting: false
}

export const isSpecialAndInterestingThunk = createAsyncThunk(
  'isSpecialAndInterestingThunk',
  () => {
    return {
      isSpecial: true,
      isInteresting: true
    }
  }
)

// This has unnecessary complexity
const loadingReducer = createReducer(initialState, builder => {
  builder.addCase(isSpecialAndInterestingThunk.fulfilled, (state, action) => {
    if (isSpecial(action)) {
      state.isSpecial = true

>     } else if (isInteresting(action)) {

Reviewer 2: Due to the `else` here, both examples actually behave differently.
In the matcher case, both execute.

Reviewer 3: Updated. I just copy and pasted that from the original and didn't bother to check 🤦

# Matching Utilities

> Redux Toolkit several type-safe utility matchers that you can leverage to simplify your code.

Reviewer 2: ```suggestion
Redux Toolkit exports several type-safe utility matchers that you can leverage to simplify your code.
```

### General Purpose

- [`isAllOf`](#isallof) - returns true when **all** conditions are met
- [`isAnyOf`](#isanyof) - returns true when at least one of the conditions are met

> ### `createAsyncThunk`-specific matchers

Reviewer 2: ```suggestion
### `createAsyncThunk`-specific matchers

All these matchers can either be called with one or more thunks as arguments, in which case they will return a matcher function for that condition and thunks, or with one actions, in which case they will match for any thunk action with said condition.
```

Accepts the same inputs as `isAllOf` and will return a type guard function that returns `true` if at least one of the provided functions match.

## `isAsyncThunkAction`

A higher-order function that returns a type guard function that may be used to check whether an action was created by [`createAsyncThunk`](./createAsyncThunk).

> ```ts title="isAsyncThunkAction usage"
> import { isAsyncThunkAction } from '@reduxjs/toolkit'
> import { request, request2 } from '@virtual/matchers'
> 
> if (isAsyncThunkAction(request, request2)) {
>   // These are all async thunk actions, handle the case.
> }
> ```

Reviewer 2: ```suggestion
```ts title="isAsyncThunkAction usage"
import { isAsyncThunkAction } from '@reduxjs/toolkit'
import { requestThunk1, requestThunk2 } from '@virtual/matchers'

const isARequestAction = isAsyncThunkAction(requestThunk1, requestThunk2);

function handleRequestAction(action: AnyAction) {
  if (isARequestAction(action)) {
    // action is either an action dispatched by `requestThunk1` or `requestThunk2`
  }
}
```
```

Reviewer 2: the same pattern would apply to the others.
